coreutils/yes.rs

```diff
@@ -37,29 +33,23 @@ use libc::FILE;
 //usage:#define yes_full_usage "\n\n"
 //usage:       "Repeatedly output a line with STRING, or 'y'"
 #[no_mangle]
-pub unsafe extern "C" fn yes_main(
-  mut _argc: libc::c_int,
-  mut argv: *mut *mut libc::c_char,
-) -> libc::c_int {
-  let mut pp: *mut *mut libc::c_char = 0 as *mut *mut libc::c_char;
-  let ref mut fresh0 = *argv.offset(0);
-  *fresh0 = b"y\x00" as *const u8 as *const libc::c_char as *mut libc::c_char;
-  if !(*argv.offset(1)).is_null() {
-    argv = argv.offset(1)
-  }
+pub unsafe extern "C" fn yes_main(mut argc: c_int, mut argv: *const *const c_char) -> c_int {
+  let line = if argc > 1 {
+    slice::from_raw_parts(argv, argc as usize)
+      .iter()
+      .map(|c_char_arg| CStr::from_ptr(*c_char_arg).to_str().unwrap())
+      .collect::<Vec<&str>>()
+      .join(" ")
+  } else {
+    "y".to_string()
+  };
+  let mut stdout = io::stdout();
   loop {
-    pp = argv;
-    loop {
-      fputs_unlocked(*pp, stdout);
-      pp = pp.offset(1);
-      if (*pp).is_null() {
-        break;
-      }
-      putchar_unlocked(' ' as i32);
-    }
-    if !(putchar_unlocked('\n' as i32) != -1i32) {
+    print!("{}", line);
+    if let Err(_) = stdout.write_all(b"\n") {
       break;
     }
   }
   bb_perror_nomsg_and_die();
```

Comment:
I think we can probably scrap this bit. Now that everything is in rust we shouldn't need any libbb error handling.

Reply:
What do you think should be done with the Rust errors, like the one, which `stdout.write_all()` returns? Should we simply print into stderr, or return the error from `main`?

Reply:
Yeah this is a good question. And this is really the first time we've had to deal with it so it's worth some consideration. Broadly speaking, I'd like to
* Maintain exit code consistency between rustybox and the corresponding tools that we are recreating. So, if `$ foo` exits with status 123 in some situation, `$ rustybox foo` should exit with status 123 in the same situation.
* Using `Result`s and idiomatic rust error handling internally is a must. When these errors must bubble up to the user we should do so in a human-friendly way -- no panic'ing.
* Panics are reserved for truly extenuating circumstances (OOM errors, etc). Prefer `.expect(...)` over `.unwrap()` for debug-ability. In most cases rustybox will distributed with the debug symbols stripped for size, and as such panics will be very hard to debug.

(These bullet points are up for discussion.)

In this case, I'm realizing now that `print!`/`println!` panic when they fail (https://doc.rust-lang.org/std/macro.println.html), so we're actually not handling all possible errors here. It looks like this is the set of bad things that can happen: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.Interrupted. The GNU coreutils version does check whether or not the write succeeds: https://github.com/coreutils/coreutils/blob/master/src/yes.c#L122 but OpenBSD's version is much simpler and doesn't include the check: https://github.com/openbsd/src/blob/master/usr.bin/yes/yes.c. Assuming both versions are functionally equivalent, I'd lean towards the simpler
```rust
loop { println!(line); }
```
version. Perhaps there's an edge case in which they differ? Wdyt?

Reply:
I have studied the implementations you have linked, plus the [Rust implementation of yes](https://github.com/uutils/coreutils/blob/master/src/yes/yes.rs). I think I am more inclined to append the `\n` symbol to the `line` and write it to the `stdout` using the `write_all` method.

The motivation for my solution is that we should be able to intercept most of the write errors and fail gracefully with the `-1` exit code, without the `println!` panics. Plus `busybox` (and `rustybox` by extension) is the kind of software to run on a restrained/embedded environment, so we should be extra careful when using something high-level like `println!`.
 
Sorry for the late reply.

P.S.: Also I agree with your bullet list for the error handling. There is nothing to change in this PR, but perhaps it could be moved to the `CONTRIBUTING.MD`

Reply:
> I have studied the implementations you have linked, plus the Rust implementation of yes. I think I am more inclined to append the \n symbol to the line and write it to the stdout using the write_all method.

So then the whole line is written at once and if that fails, we error out? That seems reasonable to me. Stdout is usually buffered on newlines anyways. Though we should probably stick to 0-255 exit codes for bash (maybe even windows) compatibility.

> Plus busybox (and rustybox by extension) is the kind of software to run on a restrained/embedded environment, so we should be extra careful when using something high-level like println!.

Well I think it's fair game to use `println!` anywhere that busybox uses `printf`, `puts`, etc. But yes, definitely important to keep in mind that `println!` can panic!

> P.S.: Also I agree with your bullet list for the error handling. There is nothing to change in this PR, but perhaps it could be moved to the CONTRIBUTING.MD?

I'll add a bit to `CONTRIBUTING.md`!

Reply:
> Though we should probably stick to 0-255 exit codes for bash (maybe even windows) compatibility

Should the error codes be declared as some high-level enum that could be accessed throughout the project?

Reply:
Oh yeah that's not a bad idea at all... It would be really nice if there was a single source of truth for what exit codes mean. I am a bit worried that there are so many utilities in here that there may be conflicts in usage though. But we could create a common type alias like `ExitCode` that is actually a `u8`.